app/src/main/java/com/youssef/fixit/Auth/LoginFragment/LoginPresenter.java

```diff
@@ -0,0 +1,108 @@
+package com.youssef.fixit.Auth.LoginFragment;
+
+import android.content.Intent;
+import android.util.Log;
+import android.widget.Toast;
+
+import com.youssef.fixit.MainActivity.MainActivity;
+import com.youssef.fixit.Models.Bids.CreateBid;
+import com.youssef.fixit.Models.Data.RetrofitClient;
+import com.youssef.fixit.Models.Register.Register;
+
+import java.util.List;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+class LoginPresenter {
+    LoginView view;
+
+    public LoginPresenter(LoginView view) {
+        this.view = view;
+    }
+
+    void OnLogin(String Mail, String Password) {
+        Validation(Mail, Password);
+    }
+
+    void Validation(String Mail, String Password) {
+        if (Mail.isEmpty()) {
+            view.OnMailIsError("Please enter your Mail");
+            return;
+        }
+        if (Password.isEmpty()) {
+            view.OnPasswordIsError("Please enter your Password");
+            return;
+        }
+
+        LoginRequest(Mail, Password);
+    }
+
+    void LoginRequest(String Mail, String Password) {
+        RetrofitClient.getInstance().Login(Mail, Password).enqueue(new Callback<Register>() {
+            @Override
+            public void onResponse(Call<Register> call, Response<Register> response) {
+                if (response.isSuccessful() && response.code() == 200) {
+                    if (response.body().getData() != null) {
+                        List<String> roles = response.body().getData().getRoles();
+                        MainActivity.MyToken = response.body().getData().getToken();
+                        MainActivity.My_ID = response.body().getData().getId();
```

Comment:
Why do you access MainActivity data using static? I think better way if you need to save `MyToken` and `MY_ID` save them in the shared preference. Static data will be lost when you're app closed and re-launch again.

Reply:
Also, always avoid using static access. this will be discussed again when we dive deep in testing and mocking so you can ignore it for now and focus on saving your data properly.

Reply:
Useing `MainActivity` as a static reference makes `LoginPresenter` tightly coupled with `MainActivity`. And also using static variable makes your app bug prone. Better save your values in shared preference and whenever you need those value, it's completely fine to fetch them from the shared preferences instead of these static fields which might get lost when you restart your app.
Also, more static fields you have the more memory you consume